File: Ruby/TddKata/KataPotter/src/book_sorter.rb
class BookSorter
  
  def sort(books)
    sets = []

    sort_in_order(books).each do |b|     
      if sets.any? { |set| set.can_add?(b) }
          add_book(sets, b)
      else
        new_set = BookSet.new
        new_set.add(b)
        sets << new_set
      end
    end
    return sets
  end

  private
  def sort_in_order(books)
    new_list = []
    books.inject(Hash.new(0)) {|h,b| h[b] += 1; h }.
          sort { |x, y| y[1] <=> x[1]}.
          each { |book, count| count.times { || new_list << book} }
    return new_list
  end

  def add_book(sets, book)
    sets.select { |s| s.can_add?(book)}.
         collect { |set| { :set => set, :cost => set.cost_with(book) } }.
         sort { |x, y| x[:cost] <=> y[:cost]}.
         first[:set].add(book)
  end

end

File: TddKatas/src/KataPotter/BookSorter.cs
using System.Collections.Generic;
using System.Linq;

namespace KataPotter
{
    public class BookSorter
    {
        public IEnumerable<BookSet> Sort(int[] books)
        {
            var booksInOrder = SortBooksInOrder(books);
            var sets = new List<BookSet>();
            booksInOrder.ForEach(b =>
            {
                if (sets.Any(s => s.CanAdd(b)))
                {
                    AddBookToSet(sets, b);
                }
                else
                {
                    sets.Add(new BookSet(b));
                }
            });
            
            return sets;
        }

        private static IEnumerable<int> SortBooksInOrder(IEnumerable<int> books)
        {
            var result = new List<int>();
            books.Distinct()
                 .Select(b => new { Book = b, Count = books.Count(book => book == b) })
                 .OrderByDescending(b => b.Count)
                 .ForEach(b => b.Count.Times(() => result.Add(b.Book)));

            return result.AsEnumerable();
        }

        private static void AddBookToSet(IEnumerable<BookSet> sets, int book)
        {
            sets.Where(s => s.CanAdd(book))
                .Select(s => new {Set = s, Cost = s.CostWith(book)})
                .OrderBy(s => s.Cost)
                .Select(s => s.Set)
                .First()
                .Add(book);
        }
    }
}